Gamma rot deadlock

RudolfWeeber opened the pull request:
Fix gamma_rot setter mpi deadlock. Add testcase, which should detect any future deadlocks in particle property setters.

RudolfWeeber committed:
Core: Fix mpi deadlock in gamma_rot and test against future deadlocks

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into python

fweik commented on testsuite/particle.py:
```diff
@@ -225,6 +224,22 @@ def get_pos_prop(handle):
+            if p in ["director","image_box","node","vs_relative"  ]: continue
+            # Uncomment to identify guilty property
+            #print( p)
+            setattr(s.part[:],p,getattr(s.part[0],p))
```
This does not work because there are read-only properties.

RudolfWeeber commented:
On Wed, May 02, 2018 at 06:41:32AM -0700, Florian Weik wrote:
> fweik commented on this pull request.
> 
> 
> 
> > @@ -225,6 +224,22 @@ def get_pos_prop(handle):
>              return handle.pos
>          self.assertRaises(RuntimeError, get_pos_prop, handle_to_non_existing_particle)
>  
> +    def test_parallel_property_setters(self):
> +        s= self.system
> +        s.part.clear()
> +        s.part.add(pos=s.box_l*np.random.random((100,3)))
> +
> +        # Copy individual properties of particle 0
> +        print("If this test hangs, there is an mpi deadlock in a particle property setter." )
> +        for p in espressomd.particle_data.particle_attributes:
> +            if p in ["director","image_box","node","vs_relative"  ]: continue
> +            # Uncomment to identify guilty property
> +            #print( p)
> +            setattr(s.part[:],p,getattr(s.part[0],p))
> 
> This does not work because there are read-only properties.
Hence the exclusions.

fweik commented:
Yes, but you are at lease missing mass. See https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/17963/raw Also this approach does not scale, it will break every time there is a new read-only property.

RudolfWeeber committed:
py: Make ParticleHandle exceptions consistent for non-writable properties

Exception -> AttributeError. AttributeError is also thrown when no property setter is defined at all.

RudolfWeeber committed:
py: fix setter for ParticleSlice.vs_relative with a single val for the entire slice

RudolfWeeber committed:
testsuite: particle property setter test handles read-only properties

Passes over AttributeErrors when setting. This is, what's thrown if a property has not setter defined.

RudolfWeeber commented:
Properties in cdef classes are implemented differently from Python properties, apparently.
So, usual ways to figure out if they have a setter don't apply.
The test now passes over AttributeErrors on setting, which is thorwn for read-only properties.
I changed those read-only properties with setters present but throwing different exception types to AttributeError.
